### PR TITLE
NIFI-12767: Error handling in Provenance and Lineage

### PR DESCRIPTION
**Pull request by mcgilman**
NIFI-12767:
- Error handling in Provenance and Lineage.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts [hunk 1]
```diff
@@ -58,10 +63,14 @@ export class ProvenanceEventListingEffects {
...
+                    catchError(() =>
                         of(
-                            ProvenanceEventListingActions.provenanceApiError({
-                                error: error.error
+                            ProvenanceEventListingActions.loadProvenanceOptionsSuccess({
+                                response: {
+                                    provenanceOptions: {
+                                        searchableFields: []
+                                    }
+                                }
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:73]: If the call to `getSearchOptions` fails, we end up with a partially complete search dialog where only the date fields are available and the buttons aren't at the bottom.
Screenshot 2024-02-13 at 1 03 18 PM

I'm not certain on what to do in this case. Does a banner error make sense? Should we just close the dialog and throw up a snackbar error?
If we can;t do anything with the dialog in the state it is in, we should probably close it and go with a snackbar error. However, It seems that you can still submit a valid query from it. So, maybe at banner error is better. We should also be sure the buttons are snapped to the bottom too.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts [hunk 2]
```diff
@@ -205,18 +211,26 @@ export class ProvenanceEventListingEffects {
...
+            tap(([, id, clusterNodeId]) => {
+                this.dialog.closeAll();
+
+                if (id) {
+                    this.provenanceService.deleteProvenanceQuery(id, clusterNodeId).subscribe({
+                        error: (errorResponse: HttpErrorResponse) => {
+                            this.store.dispatch(ErrorActions.snackBarError({ error: errorResponse.error }));
+                        }
+                    });
+                }
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:230]: Handling of the error is fine. However, what are your thoughts on not showing the error to the user in this scenario? They can't do anything about it and is pretty meaningless to them as they did get their results. It might just be more confusing to show this error.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts [hunk 3]
```diff
@@ -311,18 +321,31 @@ export class ProvenanceEventListingEffects {
...
+                                                ProvenanceEventListingActions.showOkDialog({
+                                                    title: 'Provenance',
+                                                    message: 'Successfully submitted replay request.'
+                                                })
+                                            );
+                                        },
+                                        error: (errorResponse: HttpErrorResponse) => {
+                                            this.store.dispatch(
+                                                ErrorActions.snackBarError({ error: errorResponse.error })
+                                            );
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:338]: A snackbar in this case is a bit odd. The dialog is open and the snackbar pops up well under the dialog. almost making it seem unrelated to the action they just attempted (replay of the event). Maybe a banner here?

Screenshot 2024-02-13 at 1 31 41 PM

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/provenance-event-dialog/provenance-event-dialog.component.html [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 <h2 mat-dialog-title>Provenance Event</h2>
 <div class="provenance-event">
+    <error-banner></error-banner>
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/provenance-event-dialog/provenance-event-dialog.component.html:20]: We might need to devise some way of giving these banners some context. I found a scenario where I had a banner error while on the lineage screen then opened the View Details option of one of the nodes and now the error is in 2 places at the same time. Dismissing one, dismisses them both.

Screenshot 2024-02-13 at 1 43 47 PM


Maybe we need to specify the source of the banner error in the store (optionally???) when firing them, and then the banner elements can take an input of the context they should be interested in?

```
<error-banner context="provenance-event-diaglog" ></error-banner>
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/provenance-event-table.component.html [hunk 5]
```diff
@@ -15,7 +15,8 @@
   ~ limitations under the License.
   -->
 
-<div class="provenance-event-table h-full">
+<div class="provenance-event-table h-full flex flex-col gap-y-2">
+    <error-banner></error-banner>
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/provenance-event-table.component.html:19]: As this banner is used for both provenance and lineage issues, we might need to clear them when we switch from one view to the other.

**Review by rfellows — CHANGES_REQUESTED** [review]

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts (hunk 3 above)
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:338]: I made this choice because the error isn't based on user input (there's nothing they can change and resubmit) in the dialog and a snackbar error was used in similar situations like when loading descriptors for new services and parameters.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/provenance-event-dialog/provenance-event-dialog.component.html (hunk 4 above)
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/provenance-event-dialog/provenance-event-dialog.component.html:20]: I meant to delete this `error-banner` because there was no scenario where errors could surface from the dialog directly. If I do, we could punt the contextual banner errors down the road until we hit it again.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/queue/state/queue-listing/index.ts [hunk 6]
```diff
@@ -102,7 +102,7 @@ export interface FlowFileDialogRequest {
 
 export interface QueueListingState {
     activeListingRequest: ListingRequest | null;
-    completedListingRequest: ListingRequest | null;
+    completedListingRequest: ListingRequest;
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/queue/state/queue-listing/index.ts:105]: Let me spend some time re-evaluating. Rather than being null, I provided an empty state in the provenance and lineage reducers similar to how the canvas state is initialized in the flow reducer. I should be able to make this change but may need further adjustments where this state is consumed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/lineage/index.ts [hunk 7]
```diff
@@ -66,7 +66,7 @@ export interface Lineage {
 }
 
 export interface LineageState {
-    lineage: Lineage | null;
-    error: string | null;
+    activeLineage: Lineage | null;
+    completedLineage: Lineage;
     status: 'pending' | 'loading' | 'error' | 'success';
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/lineage/index.ts:71]: I think so. It was needed in an earlier iteration of this solution. I'll remove it here and in the provenance state.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts (hunk 3 above)
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:338]: I will also update to close the dialog so it error is shown more clearly.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts (hunk 1 above)
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/state/provenance-event-listing/provenance-event-listing.effects.ts:73]: This behavior was from the current NiFi UI. The searchable fields are not loaded when the dialog opens, but rather once when the page loads. I'll update the dialog to indicate when there are no searchable fields available.

**Review by rfellows — APPROVED** [review]
Looks good. 👍  Will merge after CI completes. Thanks for all of the iterations, @mcgilman
